### DISCUSSION TITLE
Dynamically change subform field properties

### DISCUSSION THREAD

**Discussion by jameswadsworth** [joomla/joomla-cms#42834] [3. Approved Features]
While updating a component to work with Joomla 4+, [we discovered Joomla subforms](https://joomla.stackexchange.com/questions/33653/is-it-possible-to-duplicate-a-fieldset-in-joomla-form). It is great way of making 1 to many relationships and has streamlined our existing code. However an issue we encountered was that we were unable to to dynamically change the field properties for fields in the child fieldset. Well we could dynamically change the field properties for when displaying a new child fieldset but not for a child fieldset which is already populated with data.

After inspecting the code we found that the in the file /libraries/src/Form/Field/SubformField.php the function loadSubFormData when loading the data for multiple data values, creates the new form instance using the $formsource, which loads the data from the original subform field file, therefore ignoring any dynamic changes that might have taken place in a plugin. 

The new new field structure **is** available in the function loadSubFormData because it is passed in the variable Form $subForm. So changing the line 
```php 
$itemForm = Form::getInstance($subForm->getName() . $i, $this->formsource, ['control' => $control]);
```
to 
```php 
$itemForm = Form::getInstance($subForm->getName() . $i, $subForm->getXml()->saveXml(), ['control' => $control]);
```
means that the fields in the child subform now display reflecting any changes made in a plugin.

Therefore we propose making this change to the core code as being able to dynamically change field properties of the fields in the child fieldsets is a very usefully feature. It will allow developers to manipulate the field properties of the fields in the child fieldsets as we already can with other fieldset definitions. We don't think there will be any backward compatibility issues. 

Let us know what you think.

# Approval Information

Approved: 05.08.2026  
Minimum Target Version: 6.2

**Comment by Fedik** [comment]
That can work

**Reply by jameswadsworth** [reply]
I haven't understood why any new child fieldset fields show the fields as manipulated by the plugin, whereas the child fieldset fields loaded with prexisiting data don't show the changes, unless the code is changed as proposed. Both the loadSubForm() function and the loadSubFormData() function make use of the $this->formsource variable to get the variable structure. Can any one explain why?

**Reply by Fedik** [reply]
Can you post your code, how do you use it?

**Comment by jameswadsworth** [comment]
This is the code I use in my plugin to dynamically update the subform fields in a user group plugin
```php
    public function onContentPrepareForm(EventInterface $event)
    {
        [$form, $data] = array_values($event->getArguments());

        // Check we are manipulating a valid form.
        $formName = $form->getName();

        if (!\in_array($formName, ['com_advkontent.user', 'com_users.user', 'com_users.registration', 'com_users.profile'])) {
            return true;
        }

        FormHelper::addFieldPrefix('DeAnima\\Plugin\\User\\UserExtendedProfile\\Field');
        FormHelper::addFormPath(JPATH_PLUGINS . '/' . $this->_type . '/' . $this->_name . '/forms');
        $form->loadFile('userextendedprofile_multi');
        $userextendedprofileFieldset = $form->getFieldset('userextendedprofile');
        $userextendedprofileForm = $userextendedprofileFieldset['jform_userextendedprofile'];
        $uepSubform = $userextendedprofileForm->loadSubForm();
        
        $fieldInfo = [];
        foreach ($uepSubform->getFieldset('user-details') as $field)
        {
            $fieldInfo[$field->fieldname] = $field->group;
        }

        foreach ($fieldInfo as $field  => $group) {
            // Case using the advkontent users manager in admin
            if (\in_array($formName, ['com_advkontent.user', 'com_users.user', 'com_users.registration', 'com_users.profile'])) {
                if ($this->params->get('uep_required_' . $field, 1) == 0) {
                    $uepSubform->removeField($field, $group);
                } elseif ($this->params->get('uep_required_' . $field, 0) == 1) {
                    $uepSubform->setFieldAttribute($field, 'required', false, $group);
                } elseif ($this->params->get('uep_required_' . $field, 1) == 2) {
                    $uepSubform->setFieldAttribute($field, 'required', true, $group);
                }
            }
        }
        return;

    }

```
I've attached my xml form as a text file. 
[user_extended_profile.txt](https://github.com/joomla/joomla-cms/files/14421641/user_extended_profile.txt)

**Reply by Fedik** [reply]
I am sorry to say, but your code works by accident :smile: 
The reason is in `Form::getInstance()` wich return cached result, and it is deprecated already.

I not sure what can be a better way. 
Yes `$subForm->getXml()->saveXml()` will work, but not for long.

I think you can try following, for now:
```php
//... your original code, then:
$form->setFieldAttribute('userextendedprofile', 'formsource', $uepSubform->getXml()->saveXml());
```
Or probably manipulate the form XML directly.

**Comment by niosme** [comment]
So what is the proper way to manipulate subfrom fields? How to set e.g. required to false or programmatically change the variable of it? I would like here to change the order_payments fields on getForm or getFormData on the model. I think i am close enought but cant find the right way to do it.

Not the default value of it but e.g. if the amount is greater than 10 then turn required to false or something like that with the data filled.
Or if its an administrator and dont have the permission to change it to turn required to false. How we can achieve that?
Or administrator to be able to change the new added but not the previous ones data.

[order.txt](https://github.com/user-attachments/files/18767990/order.txt)
[order_payments.txt](https://github.com/user-attachments/files/18768023/order_payments.txt)
[OrderModel.txt](https://github.com/user-attachments/files/18768021/OrderModel.txt) here inside the getForm i did some tests but unsuccesfully

**Comment by HLeithner** [comment]
**The Joomla Maintainer Team decided at the 05.08.2026 meeting to accept this feature.**
